### PR TITLE
Optimize Vector4.Lerp

### PR DESCRIPTION
**Pull request by EgorBo**
### Before:
```asm
G_M10056_IG01:
       sub      rsp, 56
       vzeroupper 
       vmovaps  qword ptr [rsp+20H], xmm6
       vmovaps  qword ptr [rsp+10H], xmm7
       vmovaps  qword ptr [rsp], xmm8
G_M10056_IG02:
       vmovss   xmm0, dword ptr [rdx]
       vmovss   xmm1, dword ptr [rdx+4]
       vmovss   xmm2, dword ptr [rdx+8]
       vmovss   xmm4, dword ptr [rdx+12]
       vmovss   xmm5, dword ptr [r8]
       vmovss   xmm6, dword ptr [r8+4]
       vmovss   xmm7, dword ptr [r8+8]
       vmovss   xmm8, dword ptr [r8+12]
       vsubss   xmm5, xmm5, xmm0
       vmulss   xmm5, xmm5, xmm3
       vaddss   xmm0, xmm5, xmm0
       vsubss   xmm5, xmm6, xmm1
       vmulss   xmm5, xmm5, xmm3
       vaddss   xmm1, xmm5, xmm1
       vsubss   xmm5, xmm7, xmm2
       vmulss   xmm5, xmm5, xmm3
       vaddss   xmm2, xmm5, xmm2
       vsubss   xmm5, xmm8, xmm4
       vmulss   xmm3, xmm5, xmm3
       vaddss   xmm3, xmm3, xmm4
       vxorps   xmm4, xmm4
       vmovss   xmm4, xmm4, xmm3
       vpslldq  xmm4, 4
       vmovss   xmm4, xmm4, xmm2
       vpslldq  xmm4, 4
       vmovss   xmm4, xmm4, xmm1
       vpslldq  xmm4, 4
       vmovss   xmm4, xmm4, xmm0
       vmovaps  xmm0, xmm4
       vmovupd  xmmword ptr [rcx], xmm0
       mov      rax, rcx
G_M10056_IG03:
       vmovaps  xmm6, qword ptr [rsp+20H]
       vmovaps  xmm7, qword ptr [rsp+10H]
       vmovaps  xmm8, qword ptr [rsp]
       add      rsp, 56
       ret      
; Total bytes of code: 182
```
### After:
```asm
       vzeroupper 
G_M18874_IG02:
       vmovupd  xmm0, xmmword ptr [r8]
       vmovupd  xmm1, xmmword ptr [rdx]
       vsubps   xmm0, xmm1
       vbroadcastss xmm3, xmm3
       vmulps   xmm0, xmm3
       vaddps   xmm0, xmm1, xmm0
       vmovupd  xmmword ptr [rcx], xmm0
       mov      rax, rcx
G_M18874_IG03:
       ret
```

### CODE REVIEW

**Comment by EgorBo** [comment]
In theory, the following implementation could be better:
```csharp
private static Vector4 Lerp(Vector4 value1, Vector4 value2, float amount)
{
    Vector128<float> amountVec = Vector128.Create(amount);
    return Fma.MultiplyAdd(amountVec, value2.AsVector128(), 
        Fma.MultiplyAddNegated(amountVec, value1.AsVector128(), value1.AsVector128())).AsVector4();
}
```
but only in some sort of fast-math

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector4.cs
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/Vector4.cs:212]: Why not just `value1 + (value2 - value1) * amount`? It keeps the same ordering as we had originally (even if the eventual order of ops is still the same) and relies on the existing JIT optimization around `Vector4 * float`...

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector4.cs
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/Vector4.cs:210]: The same be done for `Vector2`/`Vector3`

Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Vector4.cs:210]: It turns out `Vector3` was already optimized.

**Comment by EgorBo** [comment]
Vector2:
### Before
```asm
       vzeroupper 
       mov      qword ptr [rsp+08H], rcx
       mov      qword ptr [rsp+10H], rdx
G_M38716_IG02:
       vmovss   xmm0, dword ptr [rsp+10H]
       vmovss   xmm1, dword ptr [rsp+08H]
       vsubss   xmm0, xmm0, xmm1
       vmulss   xmm0, xmm0, xmm2
       vaddss   xmm0, xmm0, xmm1
       vmovss   xmm1, dword ptr [rsp+14H]
       vmovss   xmm3, dword ptr [rsp+0CH]
       vsubss   xmm1, xmm1, xmm3
       vmulss   xmm1, xmm1, xmm2
       vaddss   xmm1, xmm1, xmm3
       vxorps   xmm2, xmm2
       vmovss   xmm2, xmm2, xmm1
       vpslldq  xmm2, 4
       vmovss   xmm2, xmm2, xmm0
       vmovaps  xmm0, xmm2
       vmovd    rax, xmm0
G_M38716_IG03:
       ret      
; Total bytes of code: 88
```
### After:
```asm
       push     rax
       vzeroupper 
       vmovd    xmm0, rcx
       vmovd    xmm1, rdx
G_M37838_IG02:
       vsubps   xmm1, xmm0
       vxorps   xmm3, xmm3
       vmovss   xmm3, xmm3, xmm2
       vpslldq  xmm3, 4
       vmovss   xmm3, xmm3, xmm2
       vmovaps  xmm2, xmm3
       vmulps   xmm1, xmm2
       vmovsd   qword ptr [rsp], xmm1
       vmovsd   xmm1, qword ptr [rsp]
       vaddps   xmm0, xmm1
       vmovd    rax, xmm0
G_M37838_IG03:
       add      rsp, 8
       ret      
; Total bytes of code: 67
```

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector3.cs
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/Vector3.cs]: The algorithm differs here and will cause different results in some cases due to the order of operations changing.

The difference is that in the case of `v1 + (v2 - v1) * a`, `result != v2 when a == 1`. On the other hand, `(v1 * (1 - a) ) + (v2 * a)` will guarantee it.
The former algorithm is simpler and can be used when `FMA` exists.

I'd guess this was done as a bug fix at some point (I wasn't on the team at the time and I'd have to go digging into TFS to determine the original context), so we should probably preserve the behavior.

Likewise, `Vector2/4` should probably be normalized to do the same, but it is a breaking change and we'd need sign-off (we can do that either in this pr or log an issue and follow up separately).

Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Vector3.cs]: Yes, I noticed that and reverted 🙂

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Vector3.cs]: I logged https://github.com/dotnet/runtime/issues/35529 to track normalizing all three to the same algorithm

**Comment by tannergooding** [comment]
Closing and reopening to retrigger the run against current master. It should be good to merge once tests pass.

**Review by tannergooding — APPROVED** [review]

**Comment by EgorBo** [comment]
@tannergooding the failing lane is a known issue: https://github.com/dotnet/runtime/pull/35812

**Comment by tannergooding** [comment]
Thanks! Merged.

Updating `Vector2/3/4` to be consistent is just pending final approval here at which point we can fix them up: https://github.com/dotnet/runtime/issues/35529
That will also open things up to use `System.Runtime.Intrinsics.Fma` when available.

**Pull request merged**
